In [1]:
import json
import numpy as np
import pandas as pd
import os

In [2]:
from datetime import timedelta

In [3]:
from math import sqrt
from sklearn.metrics import mean_squared_error, mean_absolute_error

In [4]:
from bokeh.plotting import figure, show, output_file, save

from bokeh.io import output_notebook

from bokeh.models import Span, Range1d

from bokeh.palettes import d3

output_notebook()

Loading BokehJS ...

In [5]:
pd.set_option('display.max_rows', 20)

In [6]:
def symmetric_mean_absolute_percentage_error(targets, predictions):
    '''
    predictions: a list with the predicted values
    targets: a list with the actual values
    '''
    import numpy as np
    # lists to NumPy arrays
    targets, predictions = np.array(targets), np.array(predictions)
    # verify predictions and targets have the same shape
    if predictions.shape == targets.shape:
            return(np.sum(2*np.abs(predictions - targets) /
                          (np.abs(targets) + np.abs(predictions)))/predictions.shape[0])

In [7]:
# build a list to select specific pickle files per model architecture
selected_pkl_list = [
    # 'DMSLSTM_TPU_006_00_test_024.pkl',
    # 'DMSLSTM_TPU_006_01_test_024.pkl',
    # 'DMSLSTM_TPU_006_02_test_024.pkl',
    # 'DMSLSTM_TPU_006_03_test_024.pkl',
    # 'DMSLSTM_TPU_006_04_test_024.pkl',
    # 'DMSLSTM_TPU_006_05_test_024.pkl',
    # 'DMSLSTM_TPU_006_06_test_024.pkl',
    # 'DMSLSTM_TPU_006_07_test_024.pkl',
    # 'DMSLSTM_TPU_006_08_test_024.pkl',
    # 'DMSLSTM_TPU_006_09_test_024.pkl',
    # 'EDSLSTM_TPU_013_00_test_024.pkl',
    # 'EDSLSTM_TPU_013_01_test_024.pkl',
    # 'EDSLSTM_TPU_013_02_test_024.pkl',
    # 'EDSLSTM_TPU_013_03_test_024.pkl',
    # 'EDSLSTM_TPU_013_04_test_024.pkl',
    # 'EDSLSTM_TPU_013_05_test_024.pkl',
    # 'EDSLSTM_TPU_013_06_test_024.pkl',
    # 'EDSLSTM_TPU_013_07_test_024.pkl',
    # 'EDSLSTM_TPU_013_08_test_024.pkl',
    # 'EDSLSTM_TPU_013_09_test_024.pkl',
    'BSCTRFM_TPU_011_00_test_024.pkl',
    'BSCTRFM_TPU_011_01_test_024.pkl',
    'BSCTRFM_TPU_011_02_test_024.pkl',
    'BSCTRFM_TPU_011_03_test_024.pkl',
    'BSCTRFM_TPU_011_04_test_024.pkl',
    'BSCTRFM_TPU_011_05_test_024.pkl',
    'BSCTRFM_TPU_011_06_test_024.pkl',
    'BSCTRFM_TPU_011_07_test_024.pkl',
    'BSCTRFM_TPU_011_08_test_024.pkl',
    'BSCTRFM_TPU_011_09_test_024.pkl',
]

In [8]:
# a Pandas dataframe to store all predictions detail items
global_df_columns = [
    'model_id', 'execution', 'dataset',
    'string_timestamps', 'predictions', 'targets',
    'mae', 'rmse', 'smape'
]

global_df = pd.DataFrame(columns=global_df_columns)

In [9]:
global_df

Empty DataFrame
Columns: [model_id, execution, dataset, string_timestamps, predictions, targets, mae, rmse, smape]
Index: []

In [10]:
# collect prediction items from all available pickle files into a single dataframe
# for available_pickle in avail_pkl_list:
#     buffer_df = pd.read_pickle('../database/predictions_detail/{}'.format(available_pickle))
#     # append buffer to final dataframe
#     df = df.append(buffer_df, ignore_index=True)  

In [11]:
# collect prediction items from all available pickle files into a single dataframe
for selected_pickle in selected_pkl_list:
    buffer_df = pd.read_pickle('../database/predictions_detail/{}'.format(selected_pickle))
    # append buffer to final dataframe
    global_df = global_df.append(buffer_df, ignore_index=True)  

In [12]:
global_df

model_id execution dataset  \
0      BSCTRFM_TPU_011         0    test   
1      BSCTRFM_TPU_011         0    test   
2      BSCTRFM_TPU_011         0    test   
3      BSCTRFM_TPU_011         0    test   
4      BSCTRFM_TPU_011         0    test   
...                ...       ...     ...   
13435  BSCTRFM_TPU_011         9    test   
13436  BSCTRFM_TPU_011         9    test   
13437  BSCTRFM_TPU_011         9    test   
13438  BSCTRFM_TPU_011         9    test   
13439  BSCTRFM_TPU_011         9    test   

                                       string_timestamps  \
0      [2018-05-14 23:00:00, 2018-05-15 00:00:00, 201...   
1      [2018-05-15 00:00:00, 2018-05-15 01:00:00, 201...   
2      [2018-05-15 01:00:00, 2018-05-15 02:00:00, 201...   
3      [2018-05-15 02:00:00, 2018-05-15 03:00:00, 201...   
4      [2018-05-15 03:00:00, 2018-05-15 04:00:00, 201...   
...                                                  ...   
13435  [2018-07-09 18:00:00, 2018-07-09 19:00:00, 201...   
13436  [2018-07-09 19:00:00, 2018-07-09 20:00:00, 201...   
13437  [2018-07-09 20:00:00, 2018-07-09 21:00:00, 201...   
13438  [2018-07-09 21:00:00, 2018-07-09 22:00:00, 201...   
13439  [2018-07-09 22:00:00, 2018-07-09 23:00:00, 201...   

                                             predictions  \
0      [2399.04248046875, 2159.216796875, 1973.577392...   
1      [2168.987548828125, 1971.5181884765625, 1859.5...   
2      [1970.384521484375, 1859.541748046875, 1789.20...   
3      [1873.4947509765625, 1774.694580078125, 1746.9...   
4      [1759.6932373046875, 1745.5762939453125, 1768....   
...                                                  ...   
13435  [2722.053466796875, 2667.195068359375, 2750.78...   
13436  [2605.48095703125, 2702.561767578125, 2748.385...   
13437  [2764.382568359375, 2800.093505859375, 2662.28...   
13438  [2842.487060546875, 2709.059814453125, 2394.05...   
13439  [2686.329833984375, 2376.633056640625, 2097.19...   

                                                 targets         mae  \
0      [2451.085, 2157.115, 1955.9133333333332, 1843....  109.123205   
1      [2157.115, 1955.9133333333332, 1843.0166666666...   98.966384   
2      [1955.9133333333332, 1843.0166666666667, 1821....  116.845018   
3      [1843.0166666666667, 1821.4583333333333, 1782....  108.792343   
4      [1821.4583333333333, 1782.45, 1804.09333333333...  120.769787   
...                                                  ...         ...   
13435  [2663.6699999999996, 2692.1733333333336, 2778....   44.559751   
13436  [2692.1733333333336, 2778.8350000000005, 2831....   63.507753   
13437  [2778.8350000000005, 2831.001666666667, 2681.7...   44.798409   
13438  [2831.001666666667, 2681.786666666667, 2392.92...   50.307819   
13439  [2681.786666666667, 2392.925, 2076.67333333333...   55.033790   

             rmse     smape inference  
0      146.486437  0.038758       024  
1      130.302648  0.035279       024  
2      150.060406  0.041870       024  
3      139.819054  0.039360       024  
4      146.576474  0.044456       024  
...           ...       ...       ...  
13435   54.611187  0.017755       024  
13436   73.075090  0.025014       024  
13437   62.312385  0.018300       024  
13438   74.629660  0.019715       024  
13439   71.972688  0.021481       024  

[13440 rows x 10 columns]

In [13]:
# day-1 predictions for this model
global_df['string_timestamps'][0]

['2018-05-14 23:00:00',
 '2018-05-15 00:00:00',
 '2018-05-15 01:00:00',
 '2018-05-15 02:00:00',
 '2018-05-15 03:00:00',
 '2018-05-15 04:00:00',
 '2018-05-15 05:00:00',
 '2018-05-15 06:00:00',
 '2018-05-15 07:00:00',
 '2018-05-15 08:00:00',
 '2018-05-15 09:00:00',
 '2018-05-15 10:00:00',
 '2018-05-15 11:00:00',
 '2018-05-15 12:00:00',
 '2018-05-15 13:00:00',
 '2018-05-15 14:00:00',
 '2018-05-15 15:00:00',
 '2018-05-15 16:00:00',
 '2018-05-15 17:00:00',
 '2018-05-15 18:00:00',
 '2018-05-15 19:00:00',
 '2018-05-15 20:00:00',
 '2018-05-15 21:00:00',
 '2018-05-15 22:00:00']

In [14]:
# day-7 predictions for this model
global_df['string_timestamps'][144]

['2018-05-20 23:00:00',
 '2018-05-21 00:00:00',
 '2018-05-21 01:00:00',
 '2018-05-21 02:00:00',
 '2018-05-21 03:00:00',
 '2018-05-21 04:00:00',
 '2018-05-21 05:00:00',
 '2018-05-21 06:00:00',
 '2018-05-21 07:00:00',
 '2018-05-21 08:00:00',
 '2018-05-21 09:00:00',
 '2018-05-21 10:00:00',
 '2018-05-21 11:00:00',
 '2018-05-21 12:00:00',
 '2018-05-21 13:00:00',
 '2018-05-21 14:00:00',
 '2018-05-21 15:00:00',
 '2018-05-21 16:00:00',
 '2018-05-21 17:00:00',
 '2018-05-21 18:00:00',
 '2018-05-21 19:00:00',
 '2018-05-21 20:00:00',
 '2018-05-21 21:00:00',
 '2018-05-21 22:00:00']

In [15]:
# a Pandas dataframe to store all predictions detail items
buffer_df_columns = [
    'timestamp',
    'model_id', 'execution',
    'prediction', 'target']

predictions_df = pd.DataFrame(columns=buffer_df_columns)

In [16]:
predictions_df

Empty DataFrame
Columns: [timestamp, model_id, execution, prediction, target]
Index: []

In [17]:
# use the prediction values for the three architectures for local forecasting comparison
# plot a number of 24-hour ahead predictions, starting the first prediction available,
# against ground thruth

In [18]:
# get a filtered dataframe from a given model-execution-dataset-inference combination
dataset, inference = 'test', '024'

for model_id in ['BSCTRFM_TPU_011']:
    for execution in [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]:
        flag = \
        global_df.model_id.eq(model_id) & \
        global_df.execution.eq(execution) & \
        global_df.dataset.eq(dataset) & \
        global_df.inference.eq(inference)
        # reset index to use the same row, drop the index column
        filtered_df = global_df[flag].reset_index(drop=True)

        # pass the array predictions and array targets for selected intervals to dataframe columns
        start_indexes = [0, 24, 48, 72, 96, 120, 144]
        for start_index in start_indexes: 

            buffer_df = pd.DataFrame(columns=buffer_df_columns)

            buffer_df['timestamp'] = pd.to_datetime(filtered_df.iloc[start_index]['string_timestamps'])
            buffer_df['model_id'] = filtered_df.iloc[start_index]['model_id']
            buffer_df['execution'] = filtered_df.iloc[start_index]['execution']
            buffer_df['prediction'] = filtered_df.iloc[start_index]['predictions']
            buffer_df['target'] = filtered_df.iloc[start_index]['targets']

            predictions_df = pd.concat([predictions_df, buffer_df])

# reset index to use the same row, drop the index column
predictions_df = predictions_df.reset_index(drop=True)

In [19]:
predictions_df

timestamp         model_id execution   prediction       target
0    2018-05-14 23:00:00  BSCTRFM_TPU_011         0  2399.042480  2451.085000
1    2018-05-15 00:00:00  BSCTRFM_TPU_011         0  2159.216797  2157.115000
2    2018-05-15 01:00:00  BSCTRFM_TPU_011         0  1973.577393  1955.913333
3    2018-05-15 02:00:00  BSCTRFM_TPU_011         0  1875.570801  1843.016667
4    2018-05-15 03:00:00  BSCTRFM_TPU_011         0  1788.225098  1821.458333
...                  ...              ...       ...          ...          ...
1675 2018-05-21 18:00:00  BSCTRFM_TPU_011         9  2834.733887  2841.735000
1676 2018-05-21 19:00:00  BSCTRFM_TPU_011         9  2757.825928  2777.203333
1677 2018-05-21 20:00:00  BSCTRFM_TPU_011         9  2786.501953  2851.368333
1678 2018-05-21 21:00:00  BSCTRFM_TPU_011         9  2882.727539  2980.845000
1679 2018-05-21 22:00:00  BSCTRFM_TPU_011         9  2734.678955  2803.260000

[1680 rows x 5 columns]

In [20]:
bigquery_arima_columns = [
    'forecast_timestamp',
    'forecast_value',
    'standard_error',
    'confidence_level',
    'prediction_interval_lower_bound',
    'prediction_interval_upper_bound',
    'confidence_interval_lower_bound',
    'confidence_interval_upper_bound'
]

In [21]:
bigquery_bsctrfm_df = pd.DataFrame(columns=bigquery_arima_columns)

In [22]:
bigquery_bsctrfm_df

Empty DataFrame
Columns: [forecast_timestamp, forecast_value, standard_error, confidence_level, prediction_interval_lower_bound, prediction_interval_upper_bound, confidence_interval_lower_bound, confidence_interval_upper_bound]
Index: []

In [23]:
csv_files = [
    'arima_bsctrfm_forecast_01.csv',
    'arima_bsctrfm_forecast_02.csv',
    'arima_bsctrfm_forecast_03.csv',
    'arima_bsctrfm_forecast_04.csv',
    'arima_bsctrfm_forecast_05.csv',
    'arima_bsctrfm_forecast_06.csv',
    'arima_bsctrfm_forecast_07.csv',
]

In [24]:
for csv_file in csv_files:
    buffer_bigquery_bsctrfm_df = pd.read_csv(
        '/home/developer/gcp/cbidmltsf/database/bigquery/{}'.format(csv_file)
    )
    bigquery_bsctrfm_df = pd.concat([bigquery_bsctrfm_df, buffer_bigquery_bsctrfm_df])


bigquery_bsctrfm_df = bigquery_bsctrfm_df.reset_index(drop=True)

In [25]:
bigquery_bsctrfm_df

forecast_timestamp  forecast_value  standard_error  confidence_level  \
0    2018-05-14T23:00:00Z     2468.502288       62.970652               0.9   
1    2018-05-15T00:00:00Z     2216.765495       72.737819               0.9   
2    2018-05-15T01:00:00Z     2017.767872       78.265392               0.9   
3    2018-05-15T02:00:00Z     1911.130583       87.725004               0.9   
4    2018-05-15T03:00:00Z     1835.874515       92.640870               0.9   
..                    ...             ...             ...               ...   
163  2018-05-21T18:00:00Z     2901.328266      119.031339               0.9   
164  2018-05-21T19:00:00Z     2846.886231      120.462841               0.9   
165  2018-05-21T20:00:00Z     2952.512151      121.877506               0.9   
166  2018-05-21T21:00:00Z     2970.495280      123.275923               0.9   
167  2018-05-21T22:00:00Z     2815.699160      124.658648               0.9   

     prediction_interval_lower_bound  prediction_interval_upper_bound  \
0                        2365.037025                      2571.967550   
1                        2097.252081                      2336.278909   
2                        1889.172263                      2146.363481   
3                        1766.992157                      2055.269009   
4                        1683.658972                      1988.090057   
..                               ...                              ...   
163                      2705.751305                      3096.905228   
164                      2648.957209                      3044.815253   
165                      2752.258734                      3152.765567   
166                      2767.944164                      3173.046396   
167                      2610.876129                      3020.522191   

     confidence_interval_lower_bound  confidence_interval_upper_bound  
0                        2365.037025                      2571.967550  
1                        2097.252081                      2336.278909  
2                        1889.172263                      2146.363481  
3                        1766.992157                      2055.269009  
4                        1683.658972                      1988.090057  
..                               ...                              ...  
163                      2705.751305                      3096.905228  
164                      2648.957209                      3044.815253  
165                      2752.258734                      3152.765567  
166                      2767.944164                      3173.046396  
167                      2610.876129                      3020.522191  

[168 rows x 8 columns]

In [26]:
# timestamp for EDSLSTM_TPU_006 first prediction is '2018-06-26 00:00:00'
# daily interval endings are '2018-06-26 23:00:00', '2018-06-27 23:00:00', ..., '2018-07-02 23:00:00'
# BigQuery ARIMA is fed with the 8 previous weeks,
# starting on '2018-05-01 00:00:00' - '2018-06-25 23:00:00' (for the first 24-hour ahead prediction)


# timestamp for EDSLSTM_TPU_013 first prediction is '2018-05-03 16:00:00'
# BigQuery ARIMA is fed with the 8 previous weeks,
# prediction for day 1 is based on a model trained from '2018-03-08 16:00:00' to '2018-05-03 15:00:00'
# prediction for day 2 is based on a model trained from '2018-03-09 16:00:00' to '2018-05-04 15:00:00'
# prediction for day 3 is based on a model trained from '2018-03-10 16:00:00' to '2018-05-05 15:00:00'
# prediction for day 4 is based on a model trained from '2018-03-11 16:00:00' to '2018-05-06 15:00:00'
# prediction for day 5 is based on a model trained from '2018-03-12 16:00:00' to '2018-05-07 15:00:00'
# prediction for day 6 is based on a model trained from '2018-03-13 16:00:00' to '2018-05-08 15:00:00'
# prediction for day 7 is based on a model trained from '2018-03-14 16:00:00' to '2018-05-09 15:00:00'


# timestamp for BSCTRFM_TPU_011 first prediction is '2018-05-14 23:00:00'
# BigQuery ARIMA is fed with the 8 previous weeks,
# prediction for day 1 is based on a model trained from '2018-03-19 23:00:00' to '2018-05-14 22:00:00'
# prediction for day 2 is based on a model trained from '2018-03-20 23:00:00' to '2018-05-15 22:00:00'
# prediction for day 3 is based on a model trained from '2018-03-21 23:00:00' to '2018-05-16 22:00:00'
# prediction for day 4 is based on a model trained from '2018-03-22 23:00:00' to '2018-05-17 22:00:00'
# prediction for day 5 is based on a model trained from '2018-03-23 23:00:00' to '2018-05-18 22:00:00'
# prediction for day 6 is based on a model trained from '2018-03-24 23:00:00' to '2018-05-19 22:00:00'
# prediction for day 7 is based on a model trained from '2018-03-25 23:00:00' to '2018-05-20 22:00:00'


In [27]:
# build separated dataframes with metrics per model and date interval

In [28]:
metrics_columns = [
    'model_id', 'execution', 'start_timestamp', 'end_timestamp', 'mae', 'rmse', 'smape'
]

In [29]:
bsctrfm_metrics_df = pd.DataFrame(columns=metrics_columns)
bsctrfm_metrics_df

Empty DataFrame
Columns: [model_id, execution, start_timestamp, end_timestamp, mae, rmse, smape]
Index: []

In [30]:
# The next cell calculates performace metrics for 1-day, 2-day, ..., up to 7-day periods
# which is not really useful as all predictions are produced on a 24-hour basis,
# then markdown the cell and change metrics calculation to a day 1, day 2, ... up to day 7 scheme

model_id = 'DMSLSTM_TPU_006'

executions = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

start_timestamp = '2018-06-26 00:00:00'

end_timestamps = [
    '2018-06-26 23:00:00',
    '2018-06-27 23:00:00',
    '2018-06-28 23:00:00',
    '2018-06-29 23:00:00',
    '2018-06-30 23:00:00',
    '2018-07-01 23:00:00',
    '2018-07-02 23:00:00',
]

for execution in executions:
    
    for end_timestamp in end_timestamps:

        flag = \
        predictions_df.model_id.eq(model_id) & \
        predictions_df.execution.eq(execution) & \
        predictions_df.timestamp.ge(start_timestamp) & \
        predictions_df.timestamp.le(end_timestamp)
        
        results_df = predictions_df[flag].reset_index(drop=True)
        
        mae = mean_absolute_error(results_df.target, results_df.prediction)
        rmse = sqrt(mean_squared_error(results_df.target, results_df.prediction))
        smape =  symmetric_mean_absolute_percentage_error(results_df.target, results_df.prediction)

        row_list = [
            model_id, execution,
            start_timestamp, end_timestamp,
            mae, rmse, smape
        ]
        row_metrics_df = pd.DataFrame([row_list], columns=metrics_columns)
        
        dmslstm_metrics_df = pd.concat([dmslstm_metrics_df, row_metrics_df])

In [31]:
# execution with best metrics over all intervals
# dmslstm_metrics_df.groupby(['execution']).mean().style.highlight_min(color = 'lightgreen', axis = 0)

In [32]:
# execution with worst metrics over all intervals
# dmslstm_metrics_df.groupby(['execution']).mean().style.highlight_max(color = 'yellow', axis = 0)

In [33]:
# metrics average for 10 executions on different intervals
# dmslstm_metrics_df.groupby(['end_timestamp']).mean()

In [34]:
# metrics standard deviation for 10 executions on different intervals
# dmslstm_metrics_df.groupby(['end_timestamp']).std()

In [35]:
# now get metrics from BigQuery ARIMA
# use target values from results_df with execution = 0 (they are the same across executions)

In [36]:
arima_one_week_mae = list()

for start_index in [0, 24, 48, 72, 96, 120, 144]:
    end_index = start_index + 24
    mae = mean_absolute_error(
        predictions_df.target[start_index:end_index],
        bigquery_bsctrfm_df.forecast_value[start_index:end_index]
    )
    
    arima_one_week_mae.append(mae)
    print('ARIMA MAE for {} to {} interval is: {:.4f}'.format(start_index, end_index, mae))

ARIMA MAE for 0 to 24 interval is: 160.6660
ARIMA MAE for 24 to 48 interval is: 100.7960
ARIMA MAE for 48 to 72 interval is: 90.7451
ARIMA MAE for 72 to 96 interval is: 71.1904
ARIMA MAE for 96 to 120 interval is: 58.2965
ARIMA MAE for 120 to 144 interval is: 124.5701
ARIMA MAE for 144 to 168 interval is: 65.9860


In [37]:
arima_one_week_rmse = list()

for start_index in [0, 24, 48, 72, 96, 120, 144]:
    end_index = start_index + 24
    rmse = sqrt(mean_squared_error(
        predictions_df.target[start_index:end_index],
        bigquery_bsctrfm_df.forecast_value[start_index:end_index]
    ))
    
    arima_one_week_rmse.append(rmse)
    print('ARIMA RMSE for {} to {} interval is: {:.4f}'.format(start_index, end_index, rmse))

ARIMA RMSE for 0 to 24 interval is: 205.4609
ARIMA RMSE for 24 to 48 interval is: 117.6361
ARIMA RMSE for 48 to 72 interval is: 104.4511
ARIMA RMSE for 72 to 96 interval is: 86.0171
ARIMA RMSE for 96 to 120 interval is: 71.5076
ARIMA RMSE for 120 to 144 interval is: 144.9715
ARIMA RMSE for 144 to 168 interval is: 75.4550


In [38]:
arima_one_week_smape = list()

for start_index in [0, 24, 48, 72, 96, 120, 144]:
    end_index = start_index + 24
    
    # adjust SMAPE to percentage value
    smape = 100*symmetric_mean_absolute_percentage_error(
        predictions_df.target[start_index:end_index],
        bigquery_bsctrfm_df.forecast_value[start_index:end_index]
    )
    
    arima_one_week_smape.append(smape)
    print('ARIMA SMAPE for {} to {} interval is: {:.4f}'.format(start_index, end_index, smape))

ARIMA SMAPE for 0 to 24 interval is: 5.8152
ARIMA SMAPE for 24 to 48 interval is: 4.3112
ARIMA SMAPE for 48 to 72 interval is: 3.6705
ARIMA SMAPE for 72 to 96 interval is: 2.8952
ARIMA SMAPE for 96 to 120 interval is: 2.4724
ARIMA SMAPE for 120 to 144 interval is: 4.8394
ARIMA SMAPE for 144 to 168 interval is: 2.7351


for end_index in [24, 48, 72, 96, 120, 144, 168]:
    mae = mean_absolute_error(
        predictions_df.target[:end_index],
        bigquery_dmslstm_df.forecast_value[:end_index]
    )
    print('ARIMA MAE for {}-hour interval is: {:.4f}'.format(end_index, mae))

for end_index in [24, 48, 72, 96, 120, 144, 168]:
    rmse = sqrt(mean_squared_error(
        predictions_df.target[:end_index],
        bigquery_dmslstm_df.forecast_value[:end_index]
    ))
    print('ARIMA RMSE for {}-hour interval is: {:.4f}'.format(end_index, rmse))

for end_index in [24, 48, 72, 96, 120, 144, 168]:
    smape = symmetric_mean_absolute_percentage_error(
        predictions_df.target[:end_index],
        bigquery_dmslstm_df.forecast_value[:end_index]
    )
    print('ARIMA SMAPE for {}-hour interval is: {:.4f}'.format(end_index, smape))

In [39]:
# a flag to isolate all pairs prediction-target for a given model in predictions_df
selected_model = 'BSCTRFM_TPU_011'
flag = predictions_df['model_id'] == selected_model

In [40]:
average_predictions_df = predictions_df[flag].groupby(['timestamp']).mean()

In [41]:
average_predictions_df

prediction       target
timestamp                                    
2018-05-14 23:00:00  2400.121509  2451.085000
2018-05-15 00:00:00  2122.198242  2157.115000
2018-05-15 01:00:00  1946.261780  1955.913333
2018-05-15 02:00:00  1839.255969  1843.016667
2018-05-15 03:00:00  1771.545740  1821.458333
...                          ...          ...
2018-05-21 18:00:00  2863.161060  2841.735000
2018-05-21 19:00:00  2804.030176  2777.203333
2018-05-21 20:00:00  2856.024292  2851.368333
2018-05-21 21:00:00  2914.492114  2980.845000
2018-05-21 22:00:00  2772.049072  2803.260000

[168 rows x 2 columns]

In [42]:
plots = dict()

In [43]:
# a datetime range for the prediction interval
# for datetime in pd.date_range(start='2018-07-02 00:00:00', end='2018-07-02 23:00:00', freq='H'):
#     print(datetime)

In [44]:
day = 7
start, end = 24*(day - 1), 24*day

size = 8

plots['bsctrfm'] = figure(
    x_axis_type='datetime',
    plot_width=960,
    plot_height=320,
    title='BSCTRFM (average) vs. ARIMA vs. Real / Day {}.'.format(day),
    toolbar_location=None,
)

plots['bsctrfm'].grid.grid_line_alpha=0.3

plots['bsctrfm'].xaxis.axis_label = 'Date'
plots['bsctrfm'].yaxis.axis_label = 'Active Power [KW]'

plots['bsctrfm'].title.text_font_size = '14pt'

plots['bsctrfm'].xaxis.axis_label_text_font_size = "14pt"
plots['bsctrfm'].yaxis.axis_label_text_font_size = "14pt"

plots['bsctrfm'].xaxis.major_label_text_font_size = "12pt"
plots['bsctrfm'].yaxis.major_label_text_font_size = "12pt"

# a custom x range to visuallly improve the the plot
plots['bsctrfm'].x_range = Range1d(
    start=average_predictions_df.index[start] - timedelta(hours=1),
    end=average_predictions_df.index[end-1] + timedelta(hours=3)
)

# the target value in average_predictions_df is the average of targets, then the only target
plots['bsctrfm'].square(
    x=average_predictions_df.index[start:end],
    y=average_predictions_df.prediction[start:end],
    size=size,
    fill_color=None,
    color='black',
    legend_label='BSCTRFM'
)
plots['bsctrfm'].line(
    x=average_predictions_df.index[start:end],
    y=average_predictions_df.prediction[start:end],
    color='black',
    legend_label='BSCTRFM'
)

plots['bsctrfm'].triangle(
    x=pd.to_datetime(bigquery_bsctrfm_df.forecast_timestamp)[start:end],
    y=bigquery_bsctrfm_df.forecast_value[start:end],
    size=size,
    fill_color=None,
    color='black',
    legend_label='ARIMA'
)
plots['bsctrfm'].line(
    x=pd.to_datetime(bigquery_bsctrfm_df.forecast_timestamp)[start:end],
    y=bigquery_bsctrfm_df.forecast_value[start:end],
    color='black',
    legend_label='ARIMA'
)

plots['bsctrfm'].circle(
    x=average_predictions_df.index[start:end],
    y=average_predictions_df.target[start:end],
    size=size,
    fill_color=None,
    color='black',
    legend_label='Real'
)
plots['bsctrfm'].line(
    x=average_predictions_df.index[start:end],
    y=average_predictions_df.target[start:end],
    color='black',
    legend_label='Real'
)

plots['bsctrfm'].legend.label_text_font_size = '12pt'
plots['bsctrfm'].legend.location = 'bottom_right'

# uncomment the following two lines to save plot
# output_file('/home/developer/gcp/cbidmltsf/datasets/cfe/{}_H_kw.html'.format(device))
# save(fig_kw)

# uncomment the following line to display plot
show(plots['bsctrfm'])

In [45]:
# a Pandas dataframe to store all predictions detail items
buffer_df_columns = [
    'start_timestamp',
    'end_timestamp',
    'model_id',
    'execution',
    'mae',
    'rmse',
    'smape'
]

day_by_day_predictions_df = pd.DataFrame(columns=buffer_df_columns)

In [46]:
day_by_day_predictions_df

Empty DataFrame
Columns: [start_timestamp, end_timestamp, model_id, execution, mae, rmse, smape]
Index: []

In [47]:

model_id, dataset, inference = 'BSCTRFM_TPU_011', 'test', '024'
executions = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

num_days = 28
start_indexes = 24*np.arange(num_days)

for execution in executions:
    
    flag = \
    global_df.model_id.eq(model_id) & \
    global_df.execution.eq(execution) & \
    global_df.dataset.eq(dataset) & \
    global_df.inference.eq(inference)

    # reset index to use the same row, drop the index column
    filtered_df = global_df[flag].reset_index(drop=True)
    
    for start_index in start_indexes:
        
        list_to_row = [
            pd.to_datetime(filtered_df.iloc[start_index]['string_timestamps'][0]),
            pd.to_datetime(filtered_df.iloc[start_index]['string_timestamps'][-1]),
            filtered_df.iloc[start_index]['model_id'],
            filtered_df.iloc[start_index]['execution'],
            filtered_df.iloc[start_index]['mae'],
            filtered_df.iloc[start_index]['rmse'],
            filtered_df.iloc[start_index]['smape']
        ]
        
        list_to_row_df = pd.DataFrame([list_to_row], columns=buffer_df_columns)

        day_by_day_predictions_df = pd.concat([day_by_day_predictions_df, list_to_row_df])

day_by_day_predictions_df = day_by_day_predictions_df.reset_index(drop=True)

In [48]:
day_by_day_predictions_df

start_timestamp       end_timestamp         model_id execution  \
0   2018-05-14 23:00:00 2018-05-15 22:00:00  BSCTRFM_TPU_011         0   
1   2018-05-15 23:00:00 2018-05-16 22:00:00  BSCTRFM_TPU_011         0   
2   2018-05-16 23:00:00 2018-05-17 22:00:00  BSCTRFM_TPU_011         0   
3   2018-05-17 23:00:00 2018-05-18 22:00:00  BSCTRFM_TPU_011         0   
4   2018-05-18 23:00:00 2018-05-19 22:00:00  BSCTRFM_TPU_011         0   
..                  ...                 ...              ...       ...   
275 2018-06-06 23:00:00 2018-06-07 22:00:00  BSCTRFM_TPU_011         9   
276 2018-06-07 23:00:00 2018-06-08 22:00:00  BSCTRFM_TPU_011         9   
277 2018-06-08 23:00:00 2018-06-09 22:00:00  BSCTRFM_TPU_011         9   
278 2018-06-09 23:00:00 2018-06-10 22:00:00  BSCTRFM_TPU_011         9   
279 2018-06-10 23:00:00 2018-06-11 22:00:00  BSCTRFM_TPU_011         9   

            mae        rmse     smape  
0    109.123205  146.486437  0.038758  
1     90.000387  115.694452  0.032139  
2    110.370862  142.735858  0.038335  
3     62.592673   73.603474  0.023556  
4    160.378849  207.757680  0.057703  
..          ...         ...       ...  
275   53.397563   72.454433  0.019158  
276  153.315137  181.741249  0.054060  
277  138.157958  162.042527  0.051972  
278   60.418738   73.855813  0.023666  
279   98.312386  110.866597  0.038712  

[280 rows x 7 columns]

In [49]:
# locating the model execution with the best performance
day_by_day_predictions_df.groupby(['execution']).mean()

mae        rmse     smape
execution                                  
0          102.593828  127.342715  0.036752
1          101.749160  126.835146  0.036303
2          109.109167  134.945207  0.039349
3          102.966932  127.123587  0.037098
4          106.292597  131.230450  0.038244
5          121.988044  148.250277  0.043944
6          105.332637  130.514264  0.037660
7           98.939150  122.640985  0.035460
8          109.267071  134.786016  0.039057
9          120.149468  144.267681  0.043571

In [50]:
# execution 7 presents the best error metrics!

In [51]:
# get the time line, on a daily basis, from the grouped dataframe
day_by_day_predictions_df.groupby(['start_timestamp']).mean().index

DatetimeIndex(['2018-05-14 23:00:00', '2018-05-15 23:00:00',
               '2018-05-16 23:00:00', '2018-05-17 23:00:00',
               '2018-05-18 23:00:00', '2018-05-19 23:00:00',
               '2018-05-20 23:00:00', '2018-05-21 23:00:00',
               '2018-05-22 23:00:00', '2018-05-23 23:00:00',
               '2018-05-24 23:00:00', '2018-05-25 23:00:00',
               '2018-05-26 23:00:00', '2018-05-27 23:00:00',
               '2018-05-28 23:00:00', '2018-05-29 23:00:00',
               '2018-05-30 23:00:00', '2018-05-31 23:00:00',
               '2018-06-01 23:00:00', '2018-06-02 23:00:00',
               '2018-06-03 23:00:00', '2018-06-04 23:00:00',
               '2018-06-05 23:00:00', '2018-06-06 23:00:00',
               '2018-06-07 23:00:00', '2018-06-08 23:00:00',
               '2018-06-09 23:00:00', '2018-06-10 23:00:00'],
              dtype='datetime64[ns]', name='start_timestamp', freq=None)

In [52]:
# compare metrics: BSCTRFM average predictions vs. ARIMA
# express them as percentage for tabulation

In [53]:
# get a metric average (on the 10 executions) for a given 24-hour interval
bsctrfm_average_one_week_mae_mean = day_by_day_predictions_df.groupby(['start_timestamp']).mean()['mae'][:7]

In [55]:
for x in bsctrfm_average_one_week_mae_mean:
    print('{:0.4f}'.format(x))

98.0120
69.7494
121.9528
65.9552
138.2748
119.1735
50.1857


In [56]:
bsctrfm_average_one_week_mae_std = day_by_day_predictions_df.groupby(['start_timestamp']).std()['mae'][:7]

In [57]:
for x in bsctrfm_average_one_week_mae_std:
    print('{:0.4f}'.format(x))

15.1866
16.6103
10.0949
11.4839
25.6004
33.1681
9.6781


In [58]:
bsctrfm_average_one_week_rmse_mean = day_by_day_predictions_df.groupby(['start_timestamp']).mean()['rmse'][:7]

In [59]:
for x in bsctrfm_average_one_week_rmse_mean:
    print('{:0.4f}'.format(x))

127.4635
89.5919
150.1624
79.5432
176.9964
140.0774
65.5064


In [60]:
bsctrfm_average_one_week_rmse_std = day_by_day_predictions_df.groupby(['start_timestamp']).std()['rmse'][:7]

In [61]:
for x in bsctrfm_average_one_week_rmse_std:
    print('{:0.4f}'.format(x))

17.8284
20.8463
13.0129
13.1284
33.3925
39.3061
14.2024


In [62]:
bsctrfm_average_one_week_smape_mean = 100*day_by_day_predictions_df.groupby(['start_timestamp']).mean()['smape'][:7]

In [63]:
for x in bsctrfm_average_one_week_smape_mean:
    print('{:0.4f}'.format(x))

3.5497
2.5155
4.3103
2.5464
4.9897
4.6401
1.8822


In [64]:
bsctrfm_average_one_week_smape_std = 100*day_by_day_predictions_df.groupby(['start_timestamp']).std()['smape'][:7]

In [65]:
for x in bsctrfm_average_one_week_smape_std:
    print('{:0.4f}'.format(x))

0.5502
0.5738
0.3610
0.4579
0.9020
1.2906
0.3484


In [66]:
baseline_mae_pct = -100*(arima_one_week_mae - bsctrfm_average_one_week_mae_mean)/arima_one_week_mae

In [67]:
for x in baseline_mae_pct:
    print('{:0.2f}'.format(x))

-39.00
-30.80
34.39
-7.35
137.19
-4.33
-23.94


In [68]:
baseline_rmse_pct = -100*(arima_one_week_rmse - bsctrfm_average_one_week_rmse_mean)/arima_one_week_rmse

In [69]:
for x in baseline_rmse_pct:
    print('{:0.2f}'.format(x))

-37.96
-23.84
43.76
-7.53
147.52
-3.38
-13.18


In [70]:
baseline_smape_pct = -100*(arima_one_week_smape - bsctrfm_average_one_week_smape_mean)/arima_one_week_smape

In [71]:
for x in baseline_smape_pct:
    print('{:0.2f}'.format(x))

-38.96
-41.65
17.43
-12.05
101.81
-4.12
-31.18


In [60]:
size = 4

plots['bsctrfm_week_metric'] = figure(
    x_axis_type='datetime',
    plot_width=960,
    plot_height=320,
    title='BSCTRFM Performance: {} for {} Consecutive 24-hour Prediction Intervals.'\
    .format('MAE', num_days),
    toolbar_location=None,
)

plots['bsctrfm_week_metric'].grid.grid_line_alpha=0.3

# a custom x range to visuallly improve the the plot
# plots['bsctrfm_week_metric'].x_range = Range1d(
#     start=pd.to_datetime('2018-05-02 16:00:00'),
#     end=pd.to_datetime('2018-06-01 16:00:00')
# )

plots['bsctrfm_week_metric'].xaxis.axis_label = 'Date'
plots['bsctrfm_week_metric'].yaxis.axis_label = 'MAE [KW]'

plots['bsctrfm_week_metric'].title.text_font_size = '14pt'

plots['bsctrfm_week_metric'].xaxis.axis_label_text_font_size = "14pt"
plots['bsctrfm_week_metric'].yaxis.axis_label_text_font_size = "14pt"

plots['bsctrfm_week_metric'].xaxis.major_label_text_font_size = "12pt"
plots['bsctrfm_week_metric'].yaxis.major_label_text_font_size = "12pt"


plots['bsctrfm_week_metric'].circle(
    x=day_by_day_predictions_df.groupby(['start_timestamp']).mean().index,
    y=day_by_day_predictions_df.groupby(['start_timestamp']).mean()['mae'],
    color='black',
    size=size,
    fill_color=None,
)
plots['bsctrfm_week_metric'].line(
    x=day_by_day_predictions_df.groupby(['start_timestamp']).mean().index,
    y=day_by_day_predictions_df.groupby(['start_timestamp']).mean()['mae'],
    color='black',
)

arima_one_week_mae_average = Span(
    location=np.mean(np.array(arima_one_week_mae)),
    dimension='width',
    line_color='black',
    line_dash='dashed',
    line_width=2)

plots['bsctrfm_week_metric'].add_layout(arima_one_week_mae_average)

show(plots['bsctrfm_week_metric'])

In [61]:
size = 4

plots['bsctrfm_week_metric'] = figure(
    x_axis_type='datetime',
    plot_width=960,
    plot_height=320,
    title='BSCTRFM Performance: {} for {} Consecutive 24-hour Prediction Intervals.'\
    .format('RMSE', num_days),
    toolbar_location=None,
)

plots['bsctrfm_week_metric'].grid.grid_line_alpha=0.3

# a custom x range to visuallly improve the the plot
# plots['bsctrfm_week_metric'].x_range = Range1d(
#     start=pd.to_datetime('2018-05-02 16:00:00'),
#     end=pd.to_datetime('2018-06-01 16:00:00')
# )

plots['bsctrfm_week_metric'].title.text_font_size = '14pt'

plots['bsctrfm_week_metric'].xaxis.axis_label_text_font_size = "14pt"
plots['bsctrfm_week_metric'].yaxis.axis_label_text_font_size = "14pt"

plots['bsctrfm_week_metric'].xaxis.major_label_text_font_size = "12pt"
plots['bsctrfm_week_metric'].yaxis.major_label_text_font_size = "12pt"

plots['bsctrfm_week_metric'].xaxis.axis_label = 'Date'
plots['bsctrfm_week_metric'].yaxis.axis_label = 'RMSE [KW]'

plots['bsctrfm_week_metric'].circle(
    x=day_by_day_predictions_df.groupby(['start_timestamp']).mean().index,
    y=day_by_day_predictions_df.groupby(['start_timestamp']).mean()['rmse'],
    color='black',
    size=size,
    fill_color=None,
)
plots['bsctrfm_week_metric'].line(
    x=day_by_day_predictions_df.groupby(['start_timestamp']).mean().index,
    y=day_by_day_predictions_df.groupby(['start_timestamp']).mean()['rmse'],
    color='black',
)

arima_one_week_rmse_average = Span(
    location=np.mean(np.array(arima_one_week_rmse)),
    dimension='width',
    line_color='black',
    line_dash='dashed',
    line_width=2)

plots['bsctrfm_week_metric'].add_layout(arima_one_week_rmse_average)


show(plots['bsctrfm_week_metric'])

In [62]:
size = 4

plots['bsctrfm_week_metric'] = figure(
    x_axis_type='datetime',
    plot_width=960,
    plot_height=320,
    title='BSCTRFM Performance: {} for {} Consecutive 24-hour Prediction Intervals.'\
    .format('SMAPE', num_days),
    toolbar_location=None,
)

plots['bsctrfm_week_metric'].grid.grid_line_alpha=0.3

# a custom x range to visuallly improve the the plot
# plots['bsctrfm_week_metric'].x_range = Range1d(
#     start=pd.to_datetime('2018-05-02 16:00:00'),
#     end=pd.to_datetime('2018-06-01 16:00:00')
# )

plots['bsctrfm_week_metric'].title.text_font_size = '14pt'

plots['bsctrfm_week_metric'].xaxis.axis_label_text_font_size = "14pt"
plots['bsctrfm_week_metric'].yaxis.axis_label_text_font_size = "14pt"

plots['bsctrfm_week_metric'].xaxis.major_label_text_font_size = "12pt"
plots['bsctrfm_week_metric'].yaxis.major_label_text_font_size = "12pt"

plots['bsctrfm_week_metric'].xaxis.axis_label = 'Date'
plots['bsctrfm_week_metric'].yaxis.axis_label = 'SMAPE [%]'

# adjust SMAPE for percentage value
plots['bsctrfm_week_metric'].circle(
    x=day_by_day_predictions_df.groupby(['start_timestamp']).mean().index,
    y=100*day_by_day_predictions_df.groupby(['start_timestamp']).mean()['smape'],
    color='black',
    size=size,
    fill_color=None,
)
# adjust SMAPE for percentage value
plots['bsctrfm_week_metric'].line(
    x=day_by_day_predictions_df.groupby(['start_timestamp']).mean().index,
    y=100*day_by_day_predictions_df.groupby(['start_timestamp']).mean()['smape'],
    color='black',
)

arima_one_week_smape_average = Span(
    location=np.mean(np.array(arima_one_week_smape)),
    dimension='width',
    line_color='black',
    line_dash='dashed',
    line_width=2)

plots['bsctrfm_week_metric'].add_layout(arima_one_week_smape_average)


show(plots['bsctrfm_week_metric'])

In [63]:
# MAE, RMSE, and SMAPE for rolling predictions in 4 weeks
# from global dataframe
start, end = 0, 672
size = 4

plots['bsctrfm_metric'] = figure(
    x_axis_type='datetime',
    plot_width=960,
    plot_height=320,
    title='BSCTRFM Performance: {} for {} Consecutive 1-hour Rolling Predictions.'\
    .format('MAE', end),
    toolbar_location=None,
)

plots['bsctrfm_metric'].grid.grid_line_alpha=0.3

plots['bsctrfm_metric'].title.text_font_size = '14pt'

plots['bsctrfm_metric'].xaxis.axis_label_text_font_size = "14pt"
plots['bsctrfm_metric'].yaxis.axis_label_text_font_size = "14pt"

plots['bsctrfm_metric'].xaxis.major_label_text_font_size = "12pt"
plots['bsctrfm_metric'].yaxis.major_label_text_font_size = "12pt"

plots['bsctrfm_metric'].xaxis.axis_label = 'Date'
plots['bsctrfm_metric'].yaxis.axis_label = 'MAE [KW]'

plots['bsctrfm_metric'].circle(
    x=[pd.to_datetime(row[0]) for row in global_df['string_timestamps'][start:end]],
    y=global_df['mae'][start:end],
    color='black',
    size=size,
    fill_color=None,
)
plots['bsctrfm_metric'].line(
    x=[pd.to_datetime(row[0]) for row in global_df['string_timestamps'][start:end]],
    y=global_df['mae'][start:end],
    color='black',
)

arima_one_week_mae_average = Span(
    location=np.mean(np.array(arima_one_week_mae)),
    dimension='width',
    line_color='black',
    line_dash='dashed',
    line_width=2)

plots['bsctrfm_metric'].add_layout(arima_one_week_mae_average)

show(plots['bsctrfm_metric'])

In [64]:
# MAE, RMSE, and SMAPE for rolling predictions in 4 weeks
# from global dataframe
start, end = 0, 672
size = 4

plots['bsctrfm_metric'] = figure(
    x_axis_type='datetime',
    plot_width=960,
    plot_height=320,
    title='BSCTRFM Performance: {} for {} Consecutive 1-hour Rolling Predictions.'\
    .format('RMSE', end),
    toolbar_location=None,
)

plots['bsctrfm_metric'].grid.grid_line_alpha=0.3

plots['bsctrfm_metric'].title.text_font_size = '14pt'

plots['bsctrfm_metric'].xaxis.axis_label_text_font_size = "14pt"
plots['bsctrfm_metric'].yaxis.axis_label_text_font_size = "14pt"

plots['bsctrfm_metric'].xaxis.major_label_text_font_size = "12pt"
plots['bsctrfm_metric'].yaxis.major_label_text_font_size = "12pt"

plots['bsctrfm_metric'].xaxis.axis_label = 'Date'
plots['bsctrfm_metric'].yaxis.axis_label = 'RMSE [KW]'

plots['bsctrfm_metric'].circle(
    x=[pd.to_datetime(row[0]) for row in global_df['string_timestamps'][start:end]],
    y=global_df['rmse'][start:end],
    color='black',
    size=size,
    fill_color=None,
)
plots['bsctrfm_metric'].line(
    x=[pd.to_datetime(row[0]) for row in global_df['string_timestamps'][start:end]],
    y=global_df['rmse'][start:end],
    color='black',
)

arima_one_week_rmse_average = Span(
    location=np.mean(np.array(arima_one_week_rmse)),
    dimension='width',
    line_color='black',
    line_dash='dashed',
    line_width=2)

plots['bsctrfm_metric'].add_layout(arima_one_week_rmse_average)

show(plots['bsctrfm_metric'])

In [65]:
# MAE, RMSE, and SMAPE for rolling predictions in 4 weeks
# from global dataframe
start, end = 0, 672
size = 4

plots['bsctrfm_metric'] = figure(
    x_axis_type='datetime',
    plot_width=960,
    plot_height=320,
    title='BSCTRFM Performance: {} for {} Consecutive 1-hour Rolling Predictions.'\
    .format('SMAPE', end),
    toolbar_location=None,
)

plots['bsctrfm_metric'].grid.grid_line_alpha=0.3

plots['bsctrfm_metric'].title.text_font_size = '14pt'

plots['bsctrfm_metric'].xaxis.axis_label_text_font_size = "14pt"
plots['bsctrfm_metric'].yaxis.axis_label_text_font_size = "14pt"

plots['bsctrfm_metric'].xaxis.major_label_text_font_size = "12pt"
plots['bsctrfm_metric'].yaxis.major_label_text_font_size = "12pt"

plots['bsctrfm_metric'].xaxis.axis_label = 'Date'
plots['bsctrfm_metric'].yaxis.axis_label = 'SMAPE [%]'

# adjust SMAPE for percentage value
plots['bsctrfm_metric'].circle(
    x=[pd.to_datetime(row[0]) for row in global_df['string_timestamps'][start:end]],
    y=100*global_df['smape'][start:end],
    color='black',
    size=size,
    fill_color=None,
)
# adjust SMAPE for percentage value
plots['bsctrfm_metric'].line(
    x=[pd.to_datetime(row[0]) for row in global_df['string_timestamps'][start:end]],
    y=100*global_df['smape'][start:end],
    color='black',
)

arima_one_week_smape_average = Span(
    location=np.mean(np.array(arima_one_week_smape)),
    dimension='width',
    line_color='black',
    line_dash='dashed',
    line_width=2)

plots['bsctrfm_metric'].add_layout(arima_one_week_smape_average)

show(plots['bsctrfm_metric'])

In [61]:
! ls -l /home/developer/DEPFIE-SCOM/ScD_Thesis/results

total 408
-rw-rw-r-- 1 developer developer    69 sep 22 13:48 BSCTRFM_TPU_010_09_eval.csv
-rw-rw-r-- 1 developer developer 46229 sep 22 13:48 BSCTRFM_TPU_010_09_loss.csv
-rw-rw-r-- 1 developer developer 46008 sep 22 13:48 BSCTRFM_TPU_010_09_lrs.csv
-rw-rw-r-- 1 developer developer    69 sep 23 17:44 BSCTRFM_TPU_011_07_eval.csv
-rw-rw-r-- 1 developer developer 45428 sep 23 17:44 BSCTRFM_TPU_011_07_loss.csv
-rw-rw-r-- 1 developer developer 45874 sep 23 17:44 BSCTRFM_TPU_011_07_lrs.csv
-rw-rw-r-- 1 developer developer 44853 sep 20 13:27 DMSLSTM_TPU_006_09_loss.csv
-rw-rw-r-- 1 developer developer 44752 sep 20 13:27 DMSLSTM_TPU_006_09_lrs.csv
-rw-rw-r-- 1 developer developer    68 sep 21 12:13 EDSLSTM_TPU_013_05_eval.csv
-rw-rw-r-- 1 developer developer 44489 sep 21 12:13 EDSLSTM_TPU_013_05_loss.csv
-rw-rw-r-- 1 developer developer 44597 sep 21 12:12 EDSLSTM_TPU_013_05_lrs.csv
-rw-rw-r-- 1 developer developer 27526 may 23 11:05 transformer_metrics.ods


In [62]:
eval_loss_df = pd.read_csv(
    '/home/developer/DEPFIE-SCOM/ScD_Thesis/results/BSCTRFM_TPU_011_07_eval.csv'
)

In [63]:
eval_loss_df

Wall time  Step     Value
0  1.627664e+09  2400  0.000305

In [64]:
loss_df = pd.read_csv(
    '/home/developer/DEPFIE-SCOM/ScD_Thesis/results/BSCTRFM_TPU_011_07_loss.csv'
)

In [65]:
loss_df

Wall time  Step     Value
0    1.627664e+09     3  0.145125
1    1.627664e+09     4  0.151897
2    1.627664e+09    10  0.052871
3    1.627664e+09    14  0.036549
4    1.627664e+09    15  0.036463
..            ...   ...       ...
995  1.627664e+09  2395  0.000319
996  1.627664e+09  2396  0.000369
997  1.627664e+09  2397  0.000379
998  1.627664e+09  2398  0.000337
999  1.627664e+09  2399  0.000364

[1000 rows x 3 columns]

In [66]:
lrs_df = pd.read_csv(
    '/home/developer/DEPFIE-SCOM/ScD_Thesis/results/BSCTRFM_TPU_011_07_lrs.csv'
)

In [67]:
lrs_df

Wall time  Step     Value
0    1.627664e+09     3  0.000010
1    1.627664e+09     4  0.000014
2    1.627664e+09    10  0.000035
3    1.627664e+09    14  0.000048
4    1.627664e+09    15  0.000052
..            ...   ...       ...
995  1.627664e+09  2395  0.000084
996  1.627664e+09  2396  0.000084
997  1.627664e+09  2397  0.000084
998  1.627664e+09  2398  0.000084
999  1.627664e+09  2399  0.000084

[1000 rows x 3 columns]

In [72]:

plots['bsctrfm_lrs'] = figure(
    plot_width=960,
    plot_height=320,
    title='BSCTRFM Learning Rate Schedule.',
    toolbar_location=None,
)

plots['bsctrfm_lrs'].grid.grid_line_alpha=0.3

# a custom x range to visuallly improve the the plot
plots['bsctrfm_lrs'].x_range = Range1d(
    start=0,
    end=2550
)

plots['bsctrfm_lrs'].title.text_font_size = '14pt'

plots['bsctrfm_lrs'].xaxis.axis_label_text_font_size = "14pt"
plots['bsctrfm_lrs'].yaxis.axis_label_text_font_size = "14pt"

plots['bsctrfm_lrs'].xaxis.major_label_text_font_size = "12pt"
plots['bsctrfm_lrs'].yaxis.major_label_text_font_size = "12pt"

plots['bsctrfm_lrs'].xaxis.axis_label = 'Training Step'
plots['bsctrfm_lrs'].yaxis.axis_label = 'Learning Rate'

# adjust SMAPE for percentage value
plots['bsctrfm_lrs'].line(
    x=lrs_df['Step'],
    y=lrs_df['Value'],
    color='black',
)

show(plots['bsctrfm_lrs'])

In [73]:
eval_loss = 0.000305

In [76]:

plots['bsctrfm_loss'] = figure(
    plot_width=960,
    plot_height=320,
    title='BSCTRFM Training Loss.',
    toolbar_location=None,
)

plots['bsctrfm_loss'].grid.grid_line_alpha=0.3

# a custom x range to visuallly improve the the plot
plots['bsctrfm_loss'].x_range = Range1d(
    start=0,
    end=2600
)

# a custom y range to visuallly improve the the plot
plots['bsctrfm_loss'].y_range = Range1d (
    start=0,
    end=0.003
)

plots['bsctrfm_loss'].title.text_font_size = '14pt'

plots['bsctrfm_loss'].xaxis.axis_label_text_font_size = "14pt"
plots['bsctrfm_loss'].yaxis.axis_label_text_font_size = "14pt"

plots['bsctrfm_loss'].xaxis.major_label_text_font_size = "12pt"
plots['bsctrfm_loss'].yaxis.major_label_text_font_size = "12pt"

plots['bsctrfm_loss'].xaxis.axis_label = 'Training Step'
plots['bsctrfm_loss'].yaxis.axis_label = 'Loss'

plots['bsctrfm_loss'].line(
    x=loss_df['Step'],
    y=loss_df['Value'],
    color='black',
)

final_eval_loss = Span(
    location=eval_loss,
    dimension='width',
    line_color='black',
    line_dash='dashed',
    line_width=2)

plots['bsctrfm_loss'].add_layout(final_eval_loss)

show(plots['bsctrfm_loss'])

In [4]:
from tensorboard.backend.event_processing import event_accumulator

In [5]:
def get_wall_time(path_to_logdir):
    '''
    receives a UNIX path to a TensorBoard logdir of a model
    returns the wall time for the model training process
    '''
    # an event accumulator to the logdir
    ea = event_accumulator.EventAccumulator(path_to_logdir,
                                            size_guidance={ # see below regarding this argument
                                                # event_accumulator.COMPRESSED_HISTOGRAMS: 500, # not used
                                                # event_accumulator.IMAGES: 4, # not used
                                                # event_accumulator.AUDIO: 4, # not used
                                                event_accumulator.SCALARS: 0, # retrieve all
                                                event_accumulator.TENSORS: 0, # retrieve all
                                                # event_accumulator.HISTOGRAMS: 1 # not used
                                            }
                                           )
    # loads events from file
    ea.Reload()
    
    # wall time is end time - start time
    wall_time = ea.Tensors('loss')[-1][0] - ea.Tensors('loss')[0][0]
    print("Wall time for model in '{}' is {} seconds.".format(path_to_logdir,
                                                            wall_time))
    return wall_time

In [6]:
models_list = [
    '/home/developer/gcp/cbidmltsf/models/BSCTRFM_TPU_011_00',
    '/home/developer/gcp/cbidmltsf/models/BSCTRFM_TPU_011_01',
    '/home/developer/gcp/cbidmltsf/models/BSCTRFM_TPU_011_02',
    '/home/developer/gcp/cbidmltsf/models/BSCTRFM_TPU_011_03',
    '/home/developer/gcp/cbidmltsf/models/BSCTRFM_TPU_011_04',
    '/home/developer/gcp/cbidmltsf/models/BSCTRFM_TPU_011_05',
    '/home/developer/gcp/cbidmltsf/models/BSCTRFM_TPU_011_06',
    '/home/developer/gcp/cbidmltsf/models/BSCTRFM_TPU_011_07',
    '/home/developer/gcp/cbidmltsf/models/BSCTRFM_TPU_011_08',
    '/home/developer/gcp/cbidmltsf/models/BSCTRFM_TPU_011_09',
]

In [7]:
# a dataframe for wall times
wall_times_list = [get_wall_time(model) for model in models_list
    
]

Found more than one graph event per run, or there was a metagraph containing a graph_def, as well as one or more graph events.  Overwriting the graph with the newest event.
Found more than one metagraph event per run. Overwriting the metagraph with the newest event.
Found more than one graph event per run, or there was a metagraph containing a graph_def, as well as one or more graph events.  Overwriting the graph with the newest event.
Found more than one metagraph event per run. Overwriting the metagraph with the newest event.


Wall time for model in '/home/developer/gcp/cbidmltsf/models/BSCTRFM_TPU_011_00' is 28.331552028656006 seconds.


Found more than one graph event per run, or there was a metagraph containing a graph_def, as well as one or more graph events.  Overwriting the graph with the newest event.
Found more than one metagraph event per run. Overwriting the metagraph with the newest event.


Wall time for model in '/home/developer/gcp/cbidmltsf/models/BSCTRFM_TPU_011_01' is 28.35389995574951 seconds.


Found more than one graph event per run, or there was a metagraph containing a graph_def, as well as one or more graph events.  Overwriting the graph with the newest event.
Found more than one metagraph event per run. Overwriting the metagraph with the newest event.


Wall time for model in '/home/developer/gcp/cbidmltsf/models/BSCTRFM_TPU_011_02' is 28.35578989982605 seconds.


Found more than one graph event per run, or there was a metagraph containing a graph_def, as well as one or more graph events.  Overwriting the graph with the newest event.
Found more than one metagraph event per run. Overwriting the metagraph with the newest event.


Wall time for model in '/home/developer/gcp/cbidmltsf/models/BSCTRFM_TPU_011_03' is 28.35229992866516 seconds.


Found more than one graph event per run, or there was a metagraph containing a graph_def, as well as one or more graph events.  Overwriting the graph with the newest event.
Found more than one metagraph event per run. Overwriting the metagraph with the newest event.


Wall time for model in '/home/developer/gcp/cbidmltsf/models/BSCTRFM_TPU_011_04' is 28.35084581375122 seconds.


Found more than one graph event per run, or there was a metagraph containing a graph_def, as well as one or more graph events.  Overwriting the graph with the newest event.
Found more than one metagraph event per run. Overwriting the metagraph with the newest event.


Wall time for model in '/home/developer/gcp/cbidmltsf/models/BSCTRFM_TPU_011_05' is 28.35304093360901 seconds.


Found more than one graph event per run, or there was a metagraph containing a graph_def, as well as one or more graph events.  Overwriting the graph with the newest event.
Found more than one metagraph event per run. Overwriting the metagraph with the newest event.


Wall time for model in '/home/developer/gcp/cbidmltsf/models/BSCTRFM_TPU_011_06' is 28.351282119750977 seconds.


Found more than one graph event per run, or there was a metagraph containing a graph_def, as well as one or more graph events.  Overwriting the graph with the newest event.
Found more than one metagraph event per run. Overwriting the metagraph with the newest event.


Wall time for model in '/home/developer/gcp/cbidmltsf/models/BSCTRFM_TPU_011_07' is 28.35225796699524 seconds.


Found more than one graph event per run, or there was a metagraph containing a graph_def, as well as one or more graph events.  Overwriting the graph with the newest event.
Found more than one metagraph event per run. Overwriting the metagraph with the newest event.


Wall time for model in '/home/developer/gcp/cbidmltsf/models/BSCTRFM_TPU_011_08' is 28.351946115493774 seconds.
Wall time for model in '/home/developer/gcp/cbidmltsf/models/BSCTRFM_TPU_011_09' is 28.351553916931152 seconds.


In [8]:
wall_times_list

[28.331552028656006,
 28.35389995574951,
 28.35578989982605,
 28.35229992866516,
 28.35084581375122,
 28.35304093360901,
 28.351282119750977,
 28.35225796699524,
 28.351946115493774,
 28.351553916931152]

In [9]:
print('{:0.4f}'.format(np.mean(wall_times_list)))

28.3504


In [10]:
print('{:0.4f}'.format(np.std(wall_times_list)))

0.0064
